packages/test/test-end-to-end-tests/src/test/containerRuntime.spec.ts

```diff
@@ -179,3 +179,74 @@ describeCompat("ContainerRuntime Document Schema", "FullCompat", (getTestObjectP
 		}
 	}
 });
+
+describeCompat("Id Compressor Schema change", "NoCompat", (getTestObjectProvider, apis) => {
+	let provider: ITestObjectProvider;
+
+	async function loadContainer(options: ITestContainerConfig) {
+		return provider.loadTestContainer(options);
+	}
+
+	async function getEntryPoint(container: IContainer) {
+		return getContainerEntryPointBackCompat<ITestDataObject>(container);
+	}
+
+	beforeEach("getTestObjectProvider", async () => {
+		provider = getTestObjectProvider();
+	});
+
+	it("upgrade", async () => {
+		const options: ITestContainerConfig = {
+			runtimeOptions: {
+				explicitSchemaControl: true,
+			},
+		};
+
+		const container = await provider.makeTestContainer({
+			runtimeOptions: {
+				explicitSchemaControl: false,
+				enableRuntimeIdCompressor: undefined,
+			},
+		});
+		const entry = await getEntryPoint(container);
+		entry._root.set("someKey", "someValue");
+
+		// ensure that old container is fully loaded (connected)
+		await provider.ensureSynchronized();
+
+		const container2 = await loadContainer({
+			runtimeOptions: {
+				explicitSchemaControl: true,
+				enableRuntimeIdCompressor: "delayed",
+			},
+		});
+		const entry2 = await getEntryPoint(container2);
+
+		// Send some ops, it will trigger schema change ops
+		// This will also trigger delay loading of ID compressor for both clients!
+		entry2._root.set("someKey2", "someValue");
+		await provider.ensureSynchronized();
+
+		// ID compressor loading is async. THere is no way to check when it's done.
```

Comment:
I wonder if you want to have some way to expose the "state" (off, loading, loaded) of ID compressor to the customer. Otherwise, they will be forced to do something like what you do now in this test.

Reply:
I do not like to do it, as it becomes public API surface that we need to support.
In general, we do not provide any guarantees how long and when we load it (and even if it's available).
I think it's more an indication that we are testing it at the wrong place - UT (as opposed to end-to-end tests) is likely better way to observe these things (working with concrete class). At the same time, I'm a bit afraid not to have end-to-end tests

Reply:
I agree with the e2e testing part. It's incredibly important to make sure the whole system works together.